sys/mathoid.js

```diff
@@ -103,6 +106,69 @@ MathoidService.prototype.checkInput = function(hyper, req) {
 };
 
 
+MathoidService.prototype._storeRenders = function(hyper, domain, hash, completeBody) {
+
+    var idx;
+    var len = FORMATS.length;
+    var reqs = new Array(len);
+
+    for (idx = 0; idx < len; idx++) {
+        var format = FORMATS[idx];
+        // ensure that we have a proper response for a given format
+        if (!completeBody[format] || !completeBody[format].headers || !completeBody[format].body) {
+            return P.reject(new HTTPError({
+                status: 504,
+                message: 'Math: missing or malformed response for format ' + format
+            }));
+        }
+        // construct the request object that will be emitted
+        var reqObj = {
+            uri: new URI([domain, 'sys', 'key_value', 'mathoid.' + format, hash]),
+            headers: Object.assign(
+                completeBody[format].headers, { 'x-resource-location': hash }),
+            body: completeBody[format].body
+        };
+        if (format === 'png' && reqObj.body && reqObj.body.type === 'Buffer') {
+            // for png, we need to convert the encoded data manually
+            // because we are receiving it wrapped inside a JSON
+            reqObj.body = new Buffer(reqObj.body.data);
+            completeBody[format].body = reqObj.body;
+        }
+        // store the emit Promise
+        reqs[idx] = hyper.put(reqObj);
+    }
+
+    // now do them all
+    return P.all(reqs).then(function() { return completeBody; });
+
+};
+
+
+MathoidService.prototype.requestAndStore = function(hyper, req) {
+
+    var self = this;
+    var rp = req.params;
+    var hash = req.headers['x-resource-location'];
```

Comment:
The client is only supplying the hash that has been previously received. If they supply an invalid hash, the request will fail because the fetch from the `post_data` bucket will fail - https://github.com/wikimedia/restbase/pull/552/files#diff-e5918fd3a44fe7806e66e07b48b078a7R148